engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java

```diff
@@ -0,0 +1,79 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2020 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+package org.pentaho.di.trans.streaming.common;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Atomic Count Semaphore
+ * <br>
+ * This class implements just a subset of the functionality of {@link Semaphore} but provides faster performance. Also,
+ * {@link Semaphore#release(int)} in testing was found not to always be thread safe.
+ *
+ * Note that this implementation has a very narrow usage scope with the following restrictions:
+ * 1. Acquire may only be called by one thread, otherwise more permits can be acquired than permitCount
+ * 2. Release should not be called more times than acquire; release is not bounded by zero and permitCount can go negative.
+ */
+public class AtomicCountSemaphore {
+
+  private Semaphore semaphore = new Semaphore( 0 );
+  private AtomicInteger permitCount;
+
+  public AtomicCountSemaphore( int count ) {
+    this.permitCount = new AtomicInteger( count );
+  }
+
+  /**
+   * Release
+   *
+   * @param count
+   * @return the permit count after the release
+   */
+  public int release( int count ) {
+    //release the semaphore if the count is more than 0 and the permit count was zero before the add
+    int afterAddCount = 0;
+    if ( count > 0 ) {
+      if ( ( afterAddCount = permitCount.getAndAdd( count ) ) == 0 ) {
+        this.semaphore.release();
+      }
```

Comment:
The above conditional is not an atomic action, so if 2 or more threads are releasing there's the possibility of semaphore.release being called extra times.  Scenario:
1) T1 `acquire()`, permitCount set to 0
2) T2 `release(1)`, sees permit count of 0, increments to 1.  Gets past line 56, but not yet to line 57.
3) T1 `acquire()`, decrements permit count to 0, `Semaphore.acquire`, blocks waiting for semaphore 
4) T3 `release(1)`, sees a permit count of 0, increments to 1, gets past line 56.
5) T2 and T3 both release the semaphore, increasing its Semaphore.permit count to 2.

This also gets a bit muddled to talk about with this class using the acquire/permit/release terminology, which is subtly different then the acquire/permit/release terminology used by the underlying semaphore.

Reply:
removed class